lib/download.gi

```diff
@@ -0,0 +1,202 @@
+##############################################################################
+##
+#W  download.gi                 GAP4 package `Utils'             Thomas Breuer
+##
+#Y  Copyright (C) 2022, The GAP Group
+
+#############################################################################
+##
+#V  Download_Methods
+##
+##  Use the following tools (in this order).
+##
+##  - If the curlInterface package is available then call its
+##    'DownloadURL' function.
+##  - If the URL starts with 'http://' and if the IO package is available
+##    then call the 'SingleHTTPRequest' function from this package.
+##  - If a 'wget' executable is available then call it.
+##  - If a 'curl' executable is available then call it.
+##
+##  Note that currently the methods are *NOT* consistent in the case of
+##  failures:
+##
+##  - The function 'SingleHTTPRequest' does not regard the error code 301
+##    as failure.
+##    This happens for example if one asks for the file at
+##    'http://www.gap-system.org/Packages/utils.html'.
+##
+BindGlobal( "Download_Methods", [] );
+
+Add( Download_Methods, rec(
+  name:= "via DownloadURL (from the curlInterface package)",
+  isAvailable:= {} -> IsPackageLoaded( "curlInterface" ),
+  download:= function( url, opt )
+    local res;
+
+    opt:= ShallowCopy( opt );
+    if not IsBound( opt.failOnError ) then
+      opt.failOnError:= true;
+    fi;
```

Comment:
Should we perhaps set default values for supported options in `Download`, and promise download methods this? That way, we ensure defauls are uniform across methods.


For `wget`, the *default* is to "fail on error"; to support `failOnError = false` there, use `--content-on-error`.